### PR TITLE
Robust control

### PR DESCRIPTION
**Pull request by nikhar-abbas**
This updates the control optimization routines to enable the use of robust controller tuning.

## Purpose
Adds the use of the robust controller tuning features available in ROSCO to WEIS. Using Level2 models, FOWT controllers can be tuned to satisfy a robust stability margin and "guarantee" stability in near-rated operation. 

## Type of change
What types of change is it?
_Select the appropriate type(s) that describe this PR_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (non-backwards-compatible fix or feature)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Documentation update
- [ ] Maintenance update
- [ ] Other (please describe)

## Testing
I added 02_control_opt/weis_driver_sm.py to the skinny tests - takes about 120s locally. 

## Checklist
_Put an `x` in the boxes that apply._

- [ ] I have run existing tests which pass locally with my changes
- [ ] I have added new tests or examples that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation

### CODE REVIEW

**Comment by nikhar-abbas** [comment]
@gbarter (or others?) I _suspect_ that these install failures are happening because of the use of numpy.distutils in the setup.py scripts - I had to move away from those in ROSCO and that seemed to solve my problems (https://github.com/NREL/ROSCO/commit/4d5adb1d9f22c4ba9712219c980fcaa00c22f462).

I'm going to make this a draft PR until @dzalkind and I get the new ROSCO version updated and merged into this - I think that should at least solve some issues.

**Comment by gbarter** [comment]
I will take a look and let you know.  The logs say the error is Cython, not distutils or the deprecation warnings for setup.py

**Comment by gbarter** [comment]
I think the CI install is fixed in #157 

When the tests pass, I will merge into `develop` and you can merge from there.

**Comment by gbarter** [comment]
Fixing the install is done and in 'develop', but there is a new problem with example 12 on linearization that I need help from @dzalkind  to address.

**Comment by nikhar-abbas** [comment]
@gbarter, thanks for this!

I did the distutils/setuptools fixes in ROSCO because I was getting similar warnings to [this](https://github.com/WISDEM/WEIS/runs/4555157426?check_suite_focus=true#step:7:11324), but somehow they were causing tests to fail in ROSCO but don't seem to be in WEIS. Should I revert the changes I made to the setup.py in this PR?

File: weis/glue_code/gc_LoadInputs.py [hunk 1]
```diff
@@ -147,6 +147,8 @@ def update_ontology_control(self, wt_opt):
         if self.modeling_options['flags']['control']:
             self.wt_init['control']['pitch']['omega_pc'] = wt_opt['tune_rosco_ivc.omega_pc']
             self.wt_init['control']['pitch']['zeta_pc']  = wt_opt['tune_rosco_ivc.zeta_pc']
+            print("wt_opt['tune_rosco_ivc.omega_pc'] = {}".format(wt_opt['tune_rosco_ivc.omega_pc']))
```
Comment (dzalkind) [inline weis/glue_code/gc_LoadInputs.py:150]: Do you need this anymore?

File: examples/12_linearization/analysis_options_doe.yaml [hunk 2]
```diff
@@ -1,5 +1,5 @@
 general:
-    folder_output: outputs/tower_doe
+    folder_output: outputs
```
Comment (dzalkind) [inline examples/12_linearization/analysis_options_doe.yaml:2]: Users will likely change this, but some standardization of outputs/ and the gitignore would be helpful.  Maybe let's revert this for now.

File: weis/glue_code/gc_LoadInputs.py (hunk 1 above)
Comment (nikhar-abbas) [reply weis/glue_code/gc_LoadInputs.py:150]: Nope, sorry about that!

File: examples/12_linearization/modeling_options_doe.yaml [hunk 3]
```diff
@@ -1,6 +1,10 @@
 General:
     verbosity: False  # When set to True, the code prints to screen many infos
-
+    openfast_configuration:
+        OF_run_fst: lin
+        OF_run_dir: outputs/openfast
+        use_exe: True
+        FAST_exe: /Users/nabbas//Documents/WindEnergyToolbox/WEIS/OpenFAST/install/bin/openfast
```
Comment (dzalkind) [inline examples/12_linearization/modeling_options_doe.yaml:7]: This is going to fail tests.

File: weis/glue_code/gc_RunTools.py [hunk 4]
```diff
@@ -20,7 +20,10 @@ def setup(self):
         self.add_input('DEL_RootMyb',   val=0.0, units = 'N*m')
         self.add_input('DEL_TwrBsMyt',  val=0.0, units = 'N*m')
         self.add_input('Std_PtfmPitch', val=0.0, units = 'deg')
-        n_PC = len(modeling_options['ROSCO']['U_pc'])
+        if modeling_options['ROSCO']['linmodel_tuning']['type'] == 'robust':
```
Comment (dzalkind) [inline weis/glue_code/gc_RunTools.py:23]: Double-check this?

**Comment by dzalkind** [comment]
I just fixed example 12 on develop.  Working on new ROSCO now.

**Comment by gbarter** [comment]
@nikhar-abbas : The `setup.py develop/install` deprecation has caught me a little flatfooted and I don't have a solid alternative for WEIS using pip.  That is one set of the warning messages.

What you referenced is something I haven't noticed before at all.  numpy.distutils previously extended the setuptools Extension to include fortran, so obviously critical to WEIS and its subtrees.  Your use of setuptools only in this PR is surprising to me that it works.  It could also be that the WEIS setup doesn't need numpy.distutils because all of the fortran compilation happens via CMake or the sub-packages (WISDEM / pyHAMS).

**Comment by dzalkind** [comment]
These tests are running locally for me, so I'm not sure what the problem is.  Have we sorted out the setup/distutils things?  I suspect there's a compilation issue because running an OpenFAST simulation causes a segfault.

**Comment by gbarter** [comment]
Let me know if I can help. . .

**Comment by nikhar-abbas** [comment]
> @nikhar-abbas : The `setup.py develop/install` deprecation has caught me a little flatfooted and I don't have a solid alternative for WEIS using pip. That is one set of the warning messages.
> 
> What you referenced is something I haven't noticed before at all. numpy.distutils previously extended the setuptools Extension to include fortran, so obviously critical to WEIS and its subtrees. Your use of setuptools only in this PR is surprising to me that it works. It could also be that the WEIS setup doesn't need numpy.distutils because all of the fortran compilation happens via CMake or the sub-packages (WISDEM / pyHAMS).

Okay, I reverted the setuptools/distutils changes for now - we can circle back to this if needed.

**Comment by nikhar-abbas** [comment]
Okay, this is tested and installing and running properly on eagle again. 

I think once these tests pass it is ready for review and to be merged

**Review by dzalkind — APPROVED** [review]

**Pull request merged**